integrals - fixed issue#19575 and now Meijerint does not fail with float input

praneethratna opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Fixes #19575 and now `Integral(0.0625*x**3.0*exp(-x/2), (x, 0, Rn)).doit()` returns 
`-0.125*Rn**3*exp(-Rn/2) - 0.75*Rn**2*exp(-Rn/2) - 3.0*Rn*exp(-Rn/2) + 6.0 - 6.0*exp(-Rn/2)` instead of 
`TypeError: 'Float' object cannot be interpreted as an integer`.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* simplify
    * fixed a bug in `hyperexpand.py` and now float cannot be interpreted as integer is not raised for some integrals.
  
<!-- END RELEASE NOTES -->

praneethratna committed:
r and k values are converted into int

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1661,14 +1661,14 @@ def try_shifted_sum(func, z):
     l = list(bbuckets[S.Zero])
     l.sort()
-    k = l[0]
+    k = int(round(l[0]))
```
`nbp.remove(k)` appears below -- isn't it possible that `k` will not be there? or are you just getting ints as floats (in which case `round` is not necessary?

smichr replied:
It also does not appear that sort is necessary -- can we just use `k = int(round(min(bbuckets[S.Zero])))` instead?

praneethratna replied:
> or are you just getting ints as floats (in which case `round` is not necessary?
Yeah, it should `Int` but I'm getting a `Float`.

praneethratna replied:
> It also does not appear that sort is necessary -- can we just use `k = int(round(min(bbuckets[S.Zero])))` instead?

Yeah, seems like we can.

praneethratna replied:
> `nbp.remove(k)` appears below -- isn't it possible that `k` will not be there? 

I don't think it is possible that `k` will not be there.

> or are you just getting ints as floats (in which case `round` is not necessary?

Yeah, it should `Int` but I'm getting a `Float`.

smichr replied:
If you don't use `round` then an error will be raised if k is not present during the `remove` attempt. You could put an inline note at that line saying that if this raises then we have to look at whether round should be used to calculate k and the raw k kept for removal purposes.

oscarbenjamin replied:
What is the value of k before this line?

praneethratna committed:
removed redundant lines

smichr committed:
Update hyperexpand.py

Since tests did not fail, the k being rounded must be equivalent to an integer or else it wouldn't have been removed successfully. So for now, assume that Float entities are integers.

The handling of b from bm at line 2334ff, however, is questionable -- gamma(float) would just give a float and this seems reasonable, so I'm not sure why the conversion to int is happening -- perhaps so the expansion below will work better?

smichr commented:
@praneethratna , see what you think of the last commit that I added.

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1661,14 +1663,12 @@ def try_shifted_sum(func, z):
     if len(abuckets[S.Zero]) != 1:
         return None
-    r = abuckets[S.Zero][0]
+    r = as_int(abuckets[S.Zero][0])  # XXX if this fails, is round necessary?
```
```suggestion
    r = as_int(abuckets[S.Zero][0], strict=False)  # XXX if this fails, is round necessary?
```

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1661,14 +1663,12 @@ def try_shifted_sum(func, z):
-    l = list(bbuckets[S.Zero])
-    l.sort()
-    k = l[0]
+    k = as_int(min(bbuckets[S.Zero]))  # XXX if this fails, is round necessary?
```
```suggestion
    k = as_int(min(bbuckets[S.Zero]), strict=False)  # XXX if this fails, is round necessary?
```

smichr committed:
use strict=False for as_int

smichr commented:
OK, look at what is being retrieved from abuckets and bbuckets: key S.Zero which means that Mod(x, 1) == 0. This will only contain integers (perhaps as floats) since the symbolic values which might give 0 are separated so if there were something like `3 + Symbol('i', integer=True)` it would be returned as `Symbol('i', integer=True)` -- only the coefficient would be removed. So `round` is not necessary.

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1661,14 +1663,12 @@ def try_shifted_sum(func, z):
     if len(abuckets[S.Zero]) != 1:
         return None
-    r = abuckets[S.Zero][0]
+    r = as_int(abuckets[S.Zero][0], strict=False)  # XXX if this fails, is round necessary?
```
```suggestion
    r = int(abuckets[S.Zero][0])
```

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -79,6 +79,8 @@
 from sympy.series import residue
 from sympy.simplify.powsimp import powdenest
 from sympy.utilities.iterables import sift
+from sympy.utilities.misc import as_int
```
```suggestion
```

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1661,14 +1663,12 @@ def try_shifted_sum(func, z):
-    l = list(bbuckets[S.Zero])
-    l.sort()
-    k = l[0]
+    k = as_int(min(bbuckets[S.Zero]), strict=False)  # XXX if this fails, is round necessary?
```
```suggestion
    k = int(min(bbuckets[S.Zero]))
```

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -2332,8 +2332,10 @@ def do_slater(an, bm, ap, bq, z, zfinal):
                 for b in bm:
-                    if not Mod(b, 1) and b.is_Number:
-                        b = int(round(b))
+                    if b.is_Number:
```
```suggestion
                    if b.is_Number:  # XXX is it necessary to convert to int?
```

smichr committed:
use int for conversion since, at worst, values are Float(Integer)

_mod1 is 0 only for numerical input x that has x % 1 == 0

smichr commented:
The only warning I would raise is that `_mod1` would be better to return `S.Zero` in the case that `Mod(x, 1) == 0` else we are using `S.Zero` to retrieve key `Float(0)` in code that does such retrieval from a sift using `_mod1`.

smichr committed:
standardize 0 output of _mod1

smichr commented:
I made the `_mod1` output for numbers be S.Zero in case `x == int(x)`.

smichr commented:
Will commit after 12h unless there is any correction needed (or sooner if anyone reviews).

smichr commented:
What do you think about this:
```diff
diff --git a/sympy/simplify/hyperexpand.py b/sympy/simplify/hyperexpand.py
index ccf0d670c3..3f92ff557e 100644
--- a/sympy/simplify/hyperexpand.py
+++ b/sympy/simplify/hyperexpand.py
@@ -83,6 +83,39 @@
 
 # function to define "buckets"
 def _mod1(x):
+    """returns additive constant, mod 1
+
+    Examples
+    ========
+
+    >>> from sympy import S, Symbol
+    >>> from sympy.simplify.hyperexpand import _mod1 as m1
+    >>> from sympy.abc import x
+
+    When the input is equal to an integer, the return value
+    will be S.Zero:
+
+    >>> m1(S(3)) is m1(S(3.0)) is S.Zero
+    True
+    >>> m1(3*S.Half)
+    1/2
+    >>> m1(S(1.5))
+    0.500000000000000
+
+    Currently, integer expressions are not reduced to 0.
+
+    >>> i = Symbol('i', integer=True)
+    >>> m1(2*i)
+    2*i
+
+    For expressions with free symbols, only the additive
+    constant is modified:
+
+    >>> m1(2*x) == m1(2*x + 3) == 2*x
+    True
+    >>> m1(2*x + S(3)/2)
+    2*x + 1/2
+    """
     # TODO see if this can work as Mod(x, 1); this will require
     # different handling of the "buckets" since these need to
     # be sorted and that fails when there is a mixture of
@@ -544,12 +577,9 @@ def build_invariants(self):
         abuckets, bbuckets = sift(self.ap, _mod1), sift(self.bq, _mod1)
 
         def tr(bucket):
-            bucket = list(bucket.items())
-            if not any(isinstance(x[0], Mod) for x in bucket):
-                bucket.sort(key=lambda x: default_sort_key(x[0]))
-            bucket = tuple([(mod, len(values)) for mod, values in bucket if
-                    values])
-            return bucket
+            items = sorted(bucket.items(), key=lambda x: x[0])
+            return tuple([(mod, len(values)) for mod, values in
+                items if values])
 
         return (self.gamma, tr(abuckets), tr(bbuckets))
 
@@ -1662,19 +1692,27 @@ def try_shifted_sum(func, z):
     abuckets, bbuckets = sift(func.ap, _mod1), sift(func.bq, _mod1)
     if len(abuckets[S.Zero]) != 1:
         return None
-    r = int(abuckets[S.Zero][0])
-    if r <= 0:
-        return None
+    else:
+        [r] = abuckets[S.Zero]
+        if r <= 0:
+            return None
     if S.Zero not in bbuckets:
         return None
-    k = int(min(bbuckets[S.Zero]))
-    if k <= 0:
-        return None
+    else:
+        k = min(bbuckets[S.Zero])
+        if k <= 0:
+            return None
 
     nap = list(func.ap)
     nap.remove(r)
     nbq = list(func.bq)
     nbq.remove(k)
+
+    # r and k might be floats equal to integers
+    # so convert them explicitly to ints now
+    r = int(r)
+    k = int(k)
+
     k -= 1
     nap = [x - k for x in nap]
     nbq = [x - k for x in nbq]
@@ -1720,7 +1758,7 @@ def try_polynomial(func, z):
     if not al0:
         return None
 
-    a = al0[-1]
+    a = int(al0[-1])  # make int in case `a` is Float equal to int
     fac = 1
     res = S.One
     for n in Tuple(*list(range(-a))):
```

praneethratna commented:
> What do you think about this:

This change is working , except i made few changes so that all tests pass.

praneethratna committed:
made changes

smichr commented on sympy/simplify/hyperexpand.py:
```diff
@@ -1720,7 +1757,7 @@ def try_polynomial(func, z):
         return None
 
-    a = al0[-1]
+    a = int(al0[-1]) # make int in case `a` is Float equal to int
```
Could you find a test that passes through here and change one or more of its numerical values to Floats to see that this works?

praneethratna replied:
I'm unable to find such a case, but is that change required? The issue is fixed even without with change.

smichr replied:
If you can't find a test that passes through there then we don't even know it is tested. And if you do find one that passes through, I'm pretty sure I can cause it to fail by replacing ints with floats.

smichr replied:
Yes...and for `r` and `k`.

smichr replied:
I set up a trap for a test passing through the polynomial case.

praneethratna replied:
> Yes...and for `r` and `k`.

The test cases are not passing, also the OP issue now returns,     
```
  File "/home/praneeth/sympy/sympy/polys/polyutils.py", line 211, in _parallel_dict_from_expr_if_gens
    monom[indices[base]] = exp
KeyError: _t**4.0
```

praneethratna committed:
made loop variable int

smichr committed:
debug

smichr committed:
test for int a in loop of try_polynomial

smichr committed:
use int only when needed

This will avoid casting solution to Integer which was presented in terms of floats.

praneethratna committed:
Update test_hyperexpand.py

praneethratna committed:
Update test_hyperexpand.py

smichr committed:
Update sympy/simplify/tests/test_hyperexpand.py

smichr commented:
Do you know if you have a test for the case when `k` is not an int?

praneethratna committed:
fixed merge conflict

praneethratna committed:
fixed quality error

smichr committed:
Update test_hyperexpand.py

smichr committed:
Update sympy/simplify/tests/test_hyperexpand.py

praneethratna committed:
Update test_hyperexpand.py

praneethratna commented:
This works for OP issue,
```diff
diff --git a/sympy/simplify/hyperexpand.py b/sympy/simplify/hyperexpand.py
index 0f10514b5b..c6f438d3ee 100644
--- a/sympy/simplify/hyperexpand.py
+++ b/sympy/simplify/hyperexpand.py
@@ -1715,7 +1715,7 @@ def try_shifted_sum(func, z):
     for n in range(int(r) - 1):
         ops.append(ShiftA(n + 1))
     ops.reverse()
-
+    k = int(k)
     fac = factorial(k)/z**k
     for a in nap:
         fac /= rf(a, k)
@@ -1725,7 +1725,7 @@ def try_shifted_sum(func, z):
     ops += [MultOperator(fac)]

     p = 0
-    for n in range(int(k)):
+    for n in range(k):
         m = z**n/factorial(n)
         for a in nap:
             m *= rf(a, n)
```
but this is giving an error,
```diff
diff --git a/sympy/simplify/hyperexpand.py b/sympy/simplify/hyperexpand.py
index 0f10514b5b..df9ce9eee7 100644
--- a/sympy/simplify/hyperexpand.py
+++ b/sympy/simplify/hyperexpand.py
@@ -1715,8 +1715,8 @@ def try_shifted_sum(func, z):
     for n in range(int(r) - 1):
         ops.append(ShiftA(n + 1))
     ops.reverse()
-
     fac = factorial(k)/z**k
+    k = int(k)
     for a in nap:
         fac /= rf(a, k)
     for b in nbq:
@@ -1725,7 +1725,7 @@ def try_shifted_sum(func, z):
     ops += [MultOperator(fac)]

     p = 0
-    for n in range(int(k)):
+    for n in range(k):
         m = z**n/factorial(n)
         for a in nap:
             m *= rf(a, n)
```

smichr commented:
> This works for OP issue,

OK, this is where we have to be careful. I don't think we should just convert to int because that can changed the output (giving an int answer instead of a float). This needs to be investigated to see what that monom is if r and k are ints  -- is it `_t` or _t**4`?

praneethratna commented:
> is it `_t` or _t**4`?

It is` _t**4` when `r` and `k` are ints.

smichr commented:
```python
>>> Poly(x**4,x)
Poly(x**4, x, domain='ZZ')
>>> Poly(x**4.,x)
Traceback (most recent call last):
```
Poly doesn't like float exp. If the returned poly would be `1.0*_t**4` then that would be ok to change the exponent to int.

oscarbenjamin commented:
It's unclear to me what any of these changes do. This needs more explanation including examples that clearly show what the values of these variables are and how the changes affect them.

praneethratna commented:
> If the returned poly would be `1.0*_t**4` then that would be ok to change the exponent to int.

So, how could we fix that here?

smichr commented:
> So, how could we fix that here?

`integrate` could raise an error if any exponent was a float. Although we can integrate `3.14*x**2 -> 3.14/3*x**3` we can't `symbolically` integrate `3.14*x**2.0` but we can numerically do so. So `Integral(3.14*x**2.0, x)` is ok (since it could be evaluated with `.evalf`). @jksuom or someone else would have to check me here.

I don't know enough about the structure of the hyperexpansion function to know what corresponds to the exponents to know where to raise errors in those routines (and at what level).

jksuom commented on sympy/simplify/hyperexpand.py:
```diff
@@ -2332,8 +2364,10 @@ def do_slater(an, bm, ap, bq, z, zfinal):
                 for b in bm:
-                    if not Mod(b, 1) and b.is_Number:
-                        b = int(round(b))
+                    if b.is_Number:  # XXX is it necessary to convert to int?
```
gamma function does not like floating point numbers that are integers in the sense that Mod(x, 1) is zero. So those should be converted to (first int's, and then at some point) Integers. Other numbers should not be converted so the Mod(b, 1) is needed.
